server/src/handler_driver.c

```diff
@@ -168,6 +203,423 @@ static bool drive_route_params_valid(const char *train_id, t_interlocking_route
+		return false;
+	}
+	if (train_id == NULL) {
+		syslog_server(LOG_ERR, "Drive-Route Decoupled: train_id is NULL");
+		return false;
+	}
+	
+	// 1. Get route signal infos and flags of segments that appear more than once
+	t_route_signal_info_array signal_infos = get_route_signal_info_array(route);
+	t_route_repeated_segment_flags repeated_segment_flags = get_route_repeated_segment_flags(route);
+	
+	// Check that signal_infos array is not empty and has at least 2 entries (source, destination)
+	if (signal_infos.data_ptr == NULL) {
+		syslog_server(LOG_ERR, "Drive-Route Decoupled: signal-infos array member null");
+		free_route_signal_info_array(&signal_infos);
+		free_route_repeated_segment_flags(&repeated_segment_flags);
+		return false;
+	} else if (signal_infos.len < 2) {
+		free_route_signal_info_array(&signal_infos);
+		free_route_repeated_segment_flags(&repeated_segment_flags);
+		syslog_server(LOG_ERR, "Drive-Route Decoupled: "
+		              "got signal-infos, but has not enough entries (< 2)");
+		return false;
+	}
+	syslog_server(LOG_DEBUG, "Drive-Route Decoupled: got signal-infos "
+	              "with arr length %d for routeID %s", signal_infos.len, route->id);
+	
+	// Signals in a route shall be set to stop once the train has driven past them.
+	// The destination signal shall not be driven past, thus there
+	// are (signal_infos.len - 1) signals to set to stop in total for driving this route
+	const size_t signals_to_set_to_stop_count = signal_infos.len - 1;
+	size_t signals_set_to_stop = 0;
+	ssize_t train_pos_index_previous = -1;
+	
+	// Main route driving loop runs as long as server is running, the drive route params are valid,
+	// and not all signals have yet been set to stop
+	while (running && drive_route_params_valid(train_id, route) 
+	       && signals_set_to_stop < signals_to_set_to_stop_count) {
+		// 1. Get position of train, and determine index (in route->path) of where the train is
+		t_train_index_on_route_query train_pos_query = get_train_pos_index_in_route_ignore_repeated_segments(train_id, route, &repeated_segment_flags);
+		
+		// 2. Check if the train_pos_query has valid pos_index
+		if (train_pos_query.err_code != OKAY_TRAIN_ON_ROUTE) {
+			syslog_server(LOG_DEBUG,
+			              "Drive-Route Decoupled: unable to determine "
+			              "the position of the train %s on the route %s", train_id, route->id);
+			// Train position unknown -> skip this iteration. 
+			// Don't want to stop route driving here as train might just be lost temporarily
+			continue;
+		}
+		
+		// 3. If train position has changed, check if any signals need to be set to aspect stop
+		if (train_pos_index_previous != (ssize_t) train_pos_query.pos_index) {
+			syslog_server(LOG_DEBUG, 
+			              "Drive-Route Decoupled: Train position index now %d",
+			              train_pos_query.pos_index);
+			if (train_pos_index_previous > (ssize_t) train_pos_query.pos_index) {
+				syslog_server(LOG_DEBUG, 
```

Comment:
This will have to be adjusted based on the next round of testing. I have not made it a warning here since for routes with segments that are traversed multiple times, the index might jump to 0 in certain valid cases. As they are valid cases, a warning would not be appropriate. However, for routes with no repeatedly occurring segments, a warning would be appropriate.    
I thought about determining whether a route has any repeatedly occurring segments at all to dynamically adjust the log level, but it would add more code to maintain here.